subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1303 @@
+{
+	evt_handler(evt);
+}
+
+static int publish_get_payload(struct mqtt_client *const client, size_t length)
+{
+	if (length > sizeof(payload_buf)) {
+		LOG_ERR("Incoming MQTT message too large for payload buffer");
+		return -EMSGSIZE;
+	}
+
+	return mqtt_readall_publish_payload(client, payload_buf, length);
+}
+
+static int topic_subscribe(void)
+{
+	int err;
+	ssize_t len;
+	char topic_devicebound[(sizeof(TOPIC_DEVICEBOUND) - 2) +
+		CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN];
+	struct mqtt_topic sub_topics[] = {
+		{
+			.topic.utf8 = topic_devicebound,
+		},
+		{
+			.topic.size = sizeof(TOPIC_TWIN_DESIRED) - 1,
+			.topic.utf8 = TOPIC_TWIN_DESIRED,
+		},
+		{
+			.topic.size = sizeof(TOPIC_TWIN_RES) - 1,
+			.topic.utf8 = TOPIC_TWIN_RES,
+		},
+		{
+			.topic.size = sizeof(TOPIC_DIRECT_METHODS) - 1,
+			.topic.utf8 = TOPIC_DIRECT_METHODS,
+		},
+	};
+	const struct mqtt_subscription_list sub_list = {
+		.list = sub_topics,
+		.list_count = ARRAY_SIZE(sub_topics),
+		.message_id = k_uptime_get_32(),
+	};
+
+	len = snprintk(topic_devicebound, sizeof(topic_devicebound),
+		       TOPIC_DEVICEBOUND, conn_config.device_id);
+	if ((len < 0) || (len > sizeof(topic_devicebound))) {
+		LOG_ERR("Failed to create device bound topic");
+		return -ENOMEM;
+	}
+
+	sub_topics[0].topic.size = len;
+
+	for (size_t i = 0; i < sub_list.list_count; i++) {
+		LOG_DBG("Subscribing to: %s",
+			log_strdup(sub_list.list[i].topic.utf8));
+	}
+
+	err = mqtt_subscribe(&client, &sub_list);
+	if (err) {
+		LOG_ERR("Failed to subscribe to topic list, error: %d", err);
+	}
+
+	LOG_DBG("Successfully subscribed to default topics");
+
+	return err;
+}
+
+static bool is_direct_method(const char *topic, size_t topic_len)
+{
+	/* Compare topic with the direct method topic prefix
+	 * `$iothub/methods/POST/`.
+	 */
+	return strncmp(TOPIC_DIRECT_METHODS, topic,
+		       MIN(sizeof(TOPIC_DIRECT_METHODS) - 2, topic_len)) == 0;
+}
+
+static bool is_device_twin_update(const char *topic, size_t topic_len)
+{
+	/* Compare topic with the device twin desired prefix
+	 * `$iothub/twin/PATCH/properties/desired/#`.
+	 */
+	return strncmp(TOPIC_TWIN_DESIRED, topic,
+		       MIN(sizeof(TOPIC_TWIN_DESIRED) - 2, topic_len)) == 0;
+}
+
+static bool is_device_twin_result(const char *topic, size_t topic_len)
+{
+	/* Compare topic with the device twin report result prefix
+	 * `$iothub/twin/res/#`.
+	 */
+	return strncmp(TOPIC_TWIN_RES, topic,
+		       MIN(sizeof(TOPIC_TWIN_RES) - 2, topic_len)) == 0;
+}
+
+/* @brief Get report result's status code.
+ *
+ * @param status Pointer to status.
+ * @param rid Pointer to request ID.
+ *
+ * @return 0 on success, or -1 on failure.
+ */
+static int get_device_twin_result(char *topic, size_t topic_len,
+				  struct azure_iot_hub_result *result)
+{
+	char *max_ptr = topic + topic_len - 1;
+	char *ptr;
+
+	/* Topic format:
+	 *   $iothub/twin/res/{status}/?$rid={request id}&$version={version}
```

Comment:
nit: alignment. Maybe it is intended, Looks good either way.

Reply:
It's intended, so I'll leave it as is